src/address.rs

```diff
@@ -145,125 +148,136 @@ impl<'s> AddressParser<'s> {
     }
 
     #[stable]
-    pub fn parse_address_list(&mut self) -> Vec<Address> {
+    pub fn parse_address_list(&mut self) -> ParsingResult<Vec<Address>> {
         let mut result = Vec::new();
+        let mut expected_separator: char;
 
         while !self.p.eof() {
             self.p.push_position();
 
-            let mut entry = self.parse_group();
-
-            if entry.is_none() {
-                // If we failed to parse as group, try again as mailbox
-                self.p.pop_position();
-
-                entry = match self.parse_mailbox() {
-                    Some(mailbox) => Some(Address::Mailbox(mailbox)),
-                    None => None,
+            match self.parse_group() {
+                Ok(x) => {
+                    // Is a group
+                    result.push(x);
+                    expected_separator = ';';
+                },
+                Err(e) => {
+                    // If we failed to parse as group, try again as mailbox
+                    self.p.pop_position();
+                    result.push(Address::Mailbox(match self.parse_mailbox() {
+                        Ok(x) => x,
+                        Err(e2) => return Err(ParsingError::new(
```

Comment:
This is a little odd due to the way Address List parsing is handled - We don't know ahead of time if it's a group or mailbox, so it perhaps doesn't make sense to say how the group parsing errored if the thing is really a Mailbox.

This whole change is also a bit unusual as to what happens with a malformed header that contains valid addresses but  is malformed at the end - would we want to say "These things were fine, others weren't"? A kind of three-state Result - OK, OK-with-Errors or Errors. This is what was stopping me from implementing Result over Option and I still don't think there's a good answer for it.

Perhaps the right way of handling this is having ParsingError types that have a field for a partial result? I'm not sure - it needs further thought.